worker/worker.go

```diff
@@ -445,6 +462,179 @@ func (w *Worker) processClaim(claim taskClaim) {
 	}
 }
 
+// superseding returns any superseding task, and a function to be called when
+// processed to resolve other superseded tasks.
+func (w *Worker) superseding(claim taskClaim) (taskClaim, func()) {
+	// Create monitor for any problems we run into
+	m := w.monitor.WithPrefix("superseding")
+
+	var payload map[string]interface{}
+	if json.Unmarshal(claim.Task.Payload, &payload) != nil {
+		// Do nothing, if there is an error, it'll show up later and logging will be
+		// more natural.
+		return claim, func() {}
+	}
+
+	// Take supersederUrl out of the payload, as it would break the payload
+	// validation done in TaskRun. We attempt to hide superseding from the rest
+	// of the worker implementation, so that it's only creating a hack here.
+	supersederURL, hasSupersederURL := payload["supersederUrl"].(string)
+	delete(payload, "supersederUrl")
+	var err error
+	claim.Task.Payload, err = json.Marshal(payload)
+	if err != nil {
+		panic(errors.Wrap(err, "failed to serialize data we know to be JSON"))
+	}
+
+	// Do nothing, if there is no supersederUrl
+	if !hasSupersederURL || supersederURL == "" {
+		return claim, func() {}
+	}
+
+	// Fetch list of superseding tasks from superseder
+	g := got.New()
+	r, err := g.Get(supersederURL + "?taskId=" + claim.Status.TaskID).Send()
+	if err != nil {
+		m.Warnf("failed to contact supersederUrl: '%s'", supersederURL)
+		return claim, func() {}
+	}
+
+	var result struct {
+		Supersedes []string `json:"supersedes"`
+	}
+	if json.Unmarshal(r.Body, &result) != nil {
+		m.Warnf("failed to JSON parse result from supersederUrl: '%s'", supersederURL)
+		return claim, func() {}
+	}
+
+	claimAttempts := make([]taskClaim, len(result.Supersedes))
+	util.Spawn(len(result.Supersedes), func(i int) {
+		taskID := result.Supersedes[i]
+		// Don't attempt to reclaim the initial task
+		if taskID == claim.Status.TaskID {
+			return
+		}
+		// Get state of the task, to find runID
+		s, qerr := w.queue.Status(taskID)
+		if qerr != nil {
+			m.WithTag("taskId", taskID).Infof("unable to get task status, error: %v", qerr)
+			return
+		}
+		runID := len(s.Status.Runs) - 1
+		if runID < 0 || s.Status.Runs[runID].State != "pending" {
+			return
+		}
+		c, qerr := w.queue.ClaimTask(taskID, strconv.Itoa(runID), &queue.TaskClaimRequest{
+			WorkerID:    w.options.WorkerID,
+			WorkerGroup: w.options.WorkerGroup,
+		})
+		if qerr != nil {
+			m.WithTags(map[string]string{
+				"taskId": taskID,
+				"runId":  strconv.Itoa(runID),
+			}).Debug("unable to claimTask from superseder, error: %v", qerr)
+			return
+		}
+		claimAttempts[i] = taskClaim(*c)
+	})
+
+	// remove invalid claims
```

Comment:
What causes an invalid claim?

Reply:
task was claimed by another worker, so it returned 409, hence, the claim object is empty... We just want those filtered out.